src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -765,15 +777,37 @@ ExecuteSingleTask(QueryDesc *queryDesc, Task *task,
 		bool queryOK = false;
 		bool failOnError = false;
 		int64 currentAffectedTupleCount = 0;
-		PGconn *connection = GetConnectionForPlacement(taskPlacement,
-													   isModificationQuery);
+		int connectionFlags = SESSION_LIFESPAN;
+		MultiConnection *multiConnection = GetNodeConnection(connectionFlags,
+															 taskPlacement->nodeName,
+															 taskPlacement->nodePort);
+		PGconn *connection = multiConnection->pgConn;
+
+		if (taskRequiresTwoPhaseCommit)
+		{
+			RemoteTransactionBeginIfNecessary(multiConnection);
+			MarkRemoteTransactionCritical(multiConnection);
+		}
 
-		if (connection == NULL)
+		if (PQstatus(multiConnection->pgConn) != CONNECTION_OK)
 		{
 			failedPlacementList = lappend(failedPlacementList, taskPlacement);
+
+			/*
+			 * If already marked as failed on InitTransactionStateForTask()
+			 * error already reported.
+			 */
+			if (!multiConnection->remoteTransaction.transactionFailed)
+			{
+				ReportConnectionError(multiConnection, WARNING);
```

Comment:
For the transaction case, we get the connection and try to send BEGIN to the worker and get the error. In the non-transaction case, since the connection is not used yet, we (should?) manually report the error. Otherwise, we mark placements INACTIVE silently.

Btw, this is the part where I'm not very confident. But, I couldn't find a better solution either.